designs/2024-deprecated-rule-metadata/README.md

```diff
@@ -0,0 +1,258 @@
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+We propose to extend `meta.deprecated` rule property schemas to reduce ambiguity and allow additional key details to be represented.
+Plugin developers are free to add additional properties (e.g. a link to the original GitHub PR).
+
+```ts
+type RuleMeta = {
+  deprecated?:
+    | boolean // Existing boolean option, backwards compatible.
+    | DeprecatedInfo // Proposed extension
+
+  /** @deprecated */
+  replacedBy?: string[] // Deprecate the top-level property and "move" into the "deprecated" object
+};
+
+/* At least one property is required */
+type DeprecatedInfo = {
+  message?: string // General message presented to the user, e.g. for the key rule why the rule is deprecated or for info how to replace the rule
+  url?: string     // URL to more information about this deprecation in general
+  replacedBy?: (string|ReplacedByInfo)[] // An empty array explicitly states that there is no replacement
+  deprecatedSince?: Version // Helps users gauge when to migrate and useful for documentation
+  availableUntil?: Version | null // The estimated version when the rule is removed (probably the next major version). null means the rule is "frozen" (will be available but will not be changed)
+}
+
+/* At least one property is required */
+type ReplacedByInfo = {
+  message?: string // General message presented to the user, e.g. how to replace the rule
+  url?: string     // URL to more information about this replacement in general
+  plugin?: Specifier // name should be "eslint" if the replacemenet is an ESLint core rule. Omit the property if the replacement is in the same plugin
+  rule?: Specifier
+}
+
+type Specifier = {
+  name?: string // Name of the plugin / rule / ...
+  url?: string  // URL to documentation for the plugin / rule
+}
+
+/* Version string of the package containing the rule (without a leading v and using the full semver, e.g. 8.53.0) */
+type Version = string
+```
+
+The `meta.replacedBy` property is moved into the `meta.deprecated` property as `meta.replacedBy` requires `meta.deprecated` to be set.
+The reason for this is that a rule logically must be marked as deprecated to be replaced by another rule which it currently can be.
+
+### Example
+Real-world example of how this could be used based on the situation in <https://github.com/eslint/eslint/issues/18053>:
+
+```js
+// lib/rules/semi.js
+module.exports = {
+  meta: {
+    deprecated: {
+      message: 'Stylistic rules are being moved out of ESLint core.',
+      url: 'https://eslint.org/blog/2023/10/deprecating-formatting-rules/',
+      replacedBy: [
+        {
+          plugin: {
+          name: '@stylistic/js',
+          url: 'https://eslint.style/rules/js/semi',
+        },
```

Comment:
This should be a URL to the plugin docs, and there should be an additional `rule` key with rule name and link to the rule docs?